### ISSUE TITLE
asyncdispatch.sendTo is unsafe

### ISSUE THREAD

**Issue by endragor** [nim-lang/Nim#4601]
From what I see in `sendTo` code, it currently accepts pointer and expects it to be valid for the whole duration of the send, which couldn't be guaranteed by the caller: GC doesn't know about the pointer and it could point to the stack memory which will no longer be valid when callbacks are invoked.
I think the right implementation would be to copy data inside `sendTo` and then refer to that copied data in callbacks (or WSA procedures).

/сс @cheatfate

**Comment by cheatfate** [comment]
what you mean with (WSA procedures)?

**Comment by endragor** [comment]
I mean Windows Sockets. Procedures like [WSASend](https://msdn.microsoft.com/en-us/library/windows/desktop/ms742203(v=vs.85).aspx) except the buffer to stay valid and untouched for the duration of the call.

**Comment by cheatfate** [comment]
After some investigation, i can say you are not fully right.
Until you are using `sendTo/recvFrom` inside of procedure marked with {.async.} macro - you are safe.
Otherwise you are right. But you can `GC_ref` your data before call to `sendTo()` and release it with `GC_unref` after call is completed. In such way you can avoid reallocation.

**Comment by endragor** [comment]
Could you elaborate why it is not always right? As I understand, you cannot make any assumptions about the pointer within the proc that accepts it. For example, one could invoke it like this:

```nim
proc sendInt32(sock: AsyncFD, i32: int32, saddr: ptr SockAddr, saddrLen: SockLen): Future[void] =
  var origI32 = i32
  var beI32: int32
  bigEndian32(addr beI32, addr origI32)
  result = socket.sendTo(addr beI32, sizeof(beI32), saddr, saddrLen)
```
`addr beI32` becomes invalid after `sendInt32` exits and its stack frame is destroyed. If I passed an addr of a heap object in the same manner, GC wouldn't know about it.

The only case when it works is when you send GCed object and the future's callback captures the reference.

So do you say the user should be aware that the procedure assumes the pointer holds valid for the duration of the call?

What do @dom96 and @Araq think?

**Comment by Araq** [comment]
> So do you say the user should be aware that the procedure assumes the pointer holds valid for the duration of the call?

That's what I think yes. It takes a pointer, length pair for absolute performance so it shouldn't allocate and copy it. We could insert a runtime check that the pointer is not pointing to the heap and we need to improve its docs.

**Comment by endragor** [comment]
Alright. But it seems like a good place to shoot yourself in the foot without knowing it until late.

Another question - `send` accepts `string` instead (and therefore is safe). Shouldn't `sendTo` and `send` be consistent in this matter?

**Comment by Araq** [comment]
What does "consistent in this matter" mean? Should the string proc be as unsafe as the pointer proc? :-)

**Comment by cheatfate** [comment]
``` nimrod
proc sendInt32(sock: AsyncFD, i32: int32, saddr: ptr SockAddr, saddrLen: SockLen): Future[void] {.async.} =
  var origI32 = i32
  var beI32: int32
  bigEndian32(addr beI32, addr origI32)
  result = await socket.sendTo(addr beI32, sizeof(beI32), saddr, saddrLen)

```

This variant is safe.

**Comment by endragor** [comment]
@Araq
I mean one of the following should hold true
- they both accept `string`
- they both accept `cstring`/`int` pair
- they both provide each option (`string` or the pair)

**Comment by cheatfate** [comment]
i vote for pointer/int pair and i can make PR. We already have asyncnet.nim which can hold `string` variants.

But this PR can break many projects...

**Comment by dom96** [comment]
The ``asyncdispatch`` module is a bit messy right now. Using the socket procedures defined in ``asyncdispatch`` is impractical and because of that I doubt they are used often.

Of these procedures, the ones which accept ``string`` should be removed. ``asyncdispatch`` should become an async ``nativesockets`` equivalent, and in the future I think it might be a good idea to rename it and move the ``Future[T]`` related functionality into some other module (I plan to write a very long issue that reviews the full stdlib and outlines what I think needs to be done with every module, so that the standard library as a whole is logical in its structure).

The lack of memory safety in these procedures is necessary to achieve maximum performance. But we should do everything we can to make them as safe as possible, this does mean that documentation needs a HUGE improvement. It also means that we shouldn't simply use the ``pointer`` type, make the type as explicit as possible (the ``cstring`` type should be used for most procedures).

I hope I have explained my thoughts well. I did just come back from work so I am pretty exhausted.

TL;DR: If you want safety then use ``asyncnet``/``net`` not ``asyncdispatch``/``nativesockets``.

**Comment by cheatfate** [comment]
` cstring` widely known as null terminated sequence. This may mislead newcomers, so i think `pointer` is much better then `cstring`, or of course we can create custom type for this procedures.

**Comment by Araq** [comment]
I prefer pointer over cstring too but we also need to keep the JS target in mind and see what can make sense with it.
